# VWL110 Ara Sınavı - Nisan 2022
--------------------


### Aşağıdaki problemlere uygun çözümler üretin.
##### Soruları kod kulanarak cevaplayın. 
##### İsterseniz çözümde bir veya daha fazla fonksiyon tanımlayabilirsiniz. Her fonksiyon için ayrı hücre açmanız gerekmemektedir.
##### Kodlama yapmadan direkt yazdığınız cevaplardan puan verilmeyecektir.
##### İnternetten kaynağını belirtmek şartıyla kaynak araştırması yapabilirsiniz, ancak arkadaşlarınızdan yardım almanız yasaktır. Tespiti halinde aşağıdaki kopya ile ilgili hükümler uygulanır.
##### Sınavınız bittikten sonra <b>isim_soyisim.ipynb</b> olarak bilgisayarınıza kaydedin ve Classroom ödevini teslim edin. Lütfen Google Colab üzerinde yaptığınız dosyayı paylaşım olarak eklemeyin.

--------------------

<br><br><br>

<b>TÜRK-ALMAN ÜNİVERSİTESİ LİSANS EĞİTİM-ÖĞRETİM
VE SINAV YÖNETMELİĞİ</b> <br>
<b>Kopya ve intihal</b> <br>
MADDE 23 – <br>(1) Bir öğrencinin sınavda kopya çekmesi, kopya vermesi, kopya çekmeye veya vermeye teşebbüs
etmesi, ödev, proje ve laboratuvar çalışmalarında kaynak belirtmeden kısmen veya tamamen intihalde bulunması veya ödev,
proje ve laboratuvar çalışmalarının kendisi tarafından yapılmadığının saptanması halinde, öğrenci hakkında Yükseköğretim
Kurumları Öğrenci Disiplin Yönetmeliği hükümlerine göre işlem yapılır.
<br>(2) Kopya çekildiği, intihalde bulunulduğu veya çalışmanın öğrencinin kendisi tarafından yapılmadığı bildirilen sınava
sıfır (0) notu verilir <br><br>
<b>YÜKSEKÖĞRETİM KURUMLARI ÖĞRENCİ DİSİPLİN YÖNETMELİĞİ</b> <br>
5/d ve 7/e maddelerine göre sınavda kopya çeken öğrenciye kınama cezası verilir ve bir yarıyıl uzaklaştırma gerektiren disiplin cezası uygulanır.
------

<br><b> Problem 1</b>. <br>(10 Puan)<br><br>x değişkeni; <br>

  <b>(a)</b> bir tam sayı, <br>
  <b>(b)</b> 13'e bölümünden 4 kalan en küçük dört basamaklı doğal sayı, <br>
  <b>(c)</b> rakamları birbirinden farklı bir sayı olduğuna göre x'i hesaplayıp yazdırın. 

In [ ]:
def myfunction():
    
    
    for i in range(1000,10000):
        if len(set(str(i)))== len(str(i)) and i%13==4:
            print(i)
            break
myfunction()  

1057


<br><br>
<b> Problem 2</b>. <br>(10 Puan)<br><br>Aşağıdaki hücreyi çalıştırın. <br>
  Buna göre metin içindeki Türkçe karakter barındıran kelimeleri listeleyin.

In [ ]:
metin ="Dünya, Güneş Sistemi'nde Güneş'e en yakın üçüncü gezegen olup şu an için üzerinde yaşam ve sıvı su barındırdığı bilinen tek astronomik cisimdir. Radyometrik tarihleme ve diğer kanıtlara göre 4,5 milyar yıldan fazla süre önce oluşmuştur. Dünya'nın yer çekimi, uzaydaki diğer nesnelerle, özellikle Güneş'le ve tek doğal uydusu Ay'la etkileşime girer. Dünya'nın Güneş'in etrafındaki yörüngesi, 365,256 güneş gün, yani bir yıldız yılı sürer. Bu süre içerisinde Dünya, kendi ekseni etrafında 366,265 kez döner."

In [ ]:
a = metin.replace(',', '').replace('.', '').split(" ")

karakterlist = ["ı", "ş", "ç", "ü", "ğ", "ö"]

#Reference: https://www.codingem.com/python-string-contains-characters/
list1 = []
for i in a:
    if (any([karakter in i for karakter in karakterlist])):
        list1.append(i)
print(list1)

['Dünya', 'Güneş', "Güneş'e", 'yakın', 'üçüncü', 'şu', 'için', 'üzerinde', 'yaşam', 'sıvı', 'barındırdığı', 'diğer', 'kanıtlara', 'göre', 'yıldan', 'süre', 'önce', 'oluşmuştur', "Dünya'nın", 'çekimi', 'diğer', 'özellikle', "Güneş'le", 'doğal', 'etkileşime', "Dünya'nın", "Güneş'in", 'etrafındaki', 'yörüngesi', 'güneş', 'gün', 'yıldız', 'yılı', 'sürer', 'süre', 'içerisinde', 'Dünya', 'etrafında', 'döner']


<br><br><b>Problem 3.</b><br>(10 Puan)<br><br>
Aşağıda verilen fonksiyon istenildiği gibi çalışmamaktadır. Sorunu tespit edip doğru çalışacak şekilde düzeltin.<br>
<b>Amaç:</b> Rakamları farklı sayılara True, olmayanlara False diye dönmek <br>
<b>Hata:</b> Kodlama hata vermiyor ama yanlış sonuç veriyor.

In [ ]:
def check_number(number):
    number = str(number)
    for num in number:
        if number.count(num) >1:
            return False
        else:
            return True

<b>Örnek:</b>

In [ ]:
check_number("1222")

True

In [ ]:
def check_number(number): 
    number = str(number)
    cnt = 0
    a = 0
    for num in number:
        cnt += 1
        if number.count(num) >1:
            a += 1
        if cnt == len(number):
            if a > 1:
                return False
            else:
                return True

print(check_number("1222"))

False


In [ ]:
check_number("1234")

True

In [ ]:
check_number("122334")

False

<br><br><b>Problem 4.</b><br>(10 Puan)<br><br>
Aşağıda verilen fonksiyon istenildiği gibi çalışmamaktadır. Sorunu tespit edip doğru çalışacak şekilde düzeltin.<br>
<b>Amaç:</b> A ile B noktaları arası 100 metredir. x değişkeni A noktasına uzaklığı metre cinsinden temsil ediyor. A'dan yola çıkan bir tavşan eğer x üçe bölünmüyorsa bir zıplamada iki metre, x üçe bölünüyorsa 6 metre gitmektedir. Amaç tavşanın A'dan B'ye kaç zıplamada (y) gideceğini bulmak.    <br>
<b>Hata:</b> Zıplama sayısı y 100 çıkıyor.

In [ ]:
x = 0
y = 0
for x in range(0,100):
    if x % 3:
        x = x+6
        y = y+1
    else:
        x = x+2
        y = y+1
print(y)

100


In [ ]:
#Problem for döngüsü içerisinde olmamızdı. Döngü her tekrar ettiğinde y bir artıyor ve sonuç 100 çıkıyordu. 
# Bu sorunu while döngüsü kullanarak giderdim.
x = 0
y = 0
while(x < 100):
    if x % 3 == 0:
        x += 6
        y += 1
    else:
        x += 2
        y += 1
print(y)

17


<br><br><b>Problem 5.</b><br>(20 Puan)<br><br>
Aşağıda verilen hataları inceleyerek hatayı düzeltin.

<br><br><b>5.a</b><br>

In [ ]:
print "1234"

SyntaxError: Missing parentheses in call to 'print'. Did you mean print("1234")? (<ipython-input-40-734611c701ff>, line 1)

In [ ]:
print("1234")

1234


<br><br><b>5.b</b><br>

In [ ]:
sayi = 1234
sayi.count(1)

AttributeError: 'int' object has no attribute 'count'

In [ ]:
sayi = 1234
str(sayi).count("1")

1

<br><br><b>5.c</b><br>

In [ ]:
aranilan_deger = "1234"
type(aranılan_deger)

NameError: name 'aranılan_deger' is not defined

In [ ]:
aranilan_deger = "1234"
type(aranilan_deger)
#hier gibt es ein Problem mit Ausdruck es muss nicht aranılan_deger,sondern aranilen_deger(ı-> i)

str

<br><br><b>5.d</b><br>

In [ ]:
for i <10:
    print(i)

SyntaxError: invalid syntax (<ipython-input-45-149f62c41b90>, line 1)

In [ ]:
i=0
while i < 10:  #richtige form
    print(i)
    i +=1


0
1
2
3
4
5
6
7
8
9


<br><br><b>Problem 6.</b><br>(20 Puan)<br><br>
Aşağıda bir sayı tahmin oyununun kodları verilmiştir. Bu oyunu istenilen şekilde revize ediniz.<br><br>
<b>Oyunun amacı:</b> Bilgisayarın tuttuğu 4 basamaklı, rakamları farklı sayıyı her tahmin sonucu bilgisayardan alınan ipuçları ile bulmak.<br>
<b>İstenilen revizyon:</b> Oyun artık sadece 4 basamaklı olmayacak, kullanıcıdan basamak sayısı istenecek ve bu isteğe göre 3-9 basamaklı bir sayı olacak. <br>

In [ ]:
import random
def basamak(sayi):
    if len(sayi)==4:
        return True
    else:
        return False
    
def rakamlarfarkli(sayi):
    sonuc = True
    for i in "0123456789":
        if sayi.count(i) > 1:
            sonuc = False
    return sonuc

def rakamsayi(sayi):
    return sayi.isdigit()

def ilkrakamsifir(sayi):
    if sayi[0] != "0":
        return True
    else:
        return False
    
def sayiuret():
    sayi = str(random.randint(1000,9999))
    while rakamlarfarkli(sayi) == False:
        sayi = str(random.randint(1000,9999))
    return sayi
    
def tahminkontrol(sayi):
    kontrol = True
    if basamak(sayi) == False:
        print("4 basamaklı bir sayı girin!")
        kontrol = False
    if rakamlarfarkli(sayi) == False:
        print("Rakamlar farklı olmalı!")
        kontrol = False
    if rakamsayi(sayi) == False:
        print("Tahmininiz rakamlardan oluşmalı!")
        kontrol = False
    if ilkrakamsifir(sayi) == False:
        print("İlk rakam 0 olamaz!")
        kontrol = False
    return kontrol

def skorla(sayi, tahmin):
    artilar = 0
    eksiler = 0

    for i in tahmin:
        if i in sayi:
            if sayi.index(i) == tahmin.index(i):
                artilar += 1
            else:
                eksiler += 1
    if artilar == 4:
        print("Kazandın!")
    else:
        print(" + ",artilar,"\n","- ",eksiler)
    return artilar

def degerlendir(i):
    if i <4:
        print("Şanlısın!")
    elif i <6:
        print("Dahi olmalısın!")
    elif i<7:
        print("Sen bu işi biliyorsun!")
    elif i<9:
        print("Fena değil!")
    elif i<11:
        print("Biraz kafa yorman gerek!")
    else:
        print("Sallaya sallaya tutturdun sonunda!")
    

def oyun():
    skor=0
    i=0
    sayi = sayiuret()
    while skor != 4:
        kontrol = False
        while kontrol == False:
            tahmin = input("\nSayı Girin: ")
            kontrol = tahminkontrol(tahmin)
        skor = skorla(sayi,tahmin)
        i += 1
    degerlendir(i)
    print(i,"denemede kazandınız!")

In [ ]:
import random

inp = int(input("Kaç basamaklı bir sayıyı tahmin etmek istiyorsun (3-9 arası): "))

def basamak(sayi):
    if len(sayi)==inp:
        return True
    else:
        return False

def rakamlarfarkli(sayi):
    sonuc = True
    for i in "0123456789":
        if sayi.count(i) > 1:
            sonuc = False
    return sonuc

def rakamsayi(sayi):
    return sayi.isdigit()

def ilkrakamsifir(sayi):
    if sayi[0] != "0":
        return True
    else:
        return False
    
def sayiuret(n):
    sayi = str(random.randint(pow(10,n-1), pow(10,n) - 1))
    while rakamlarfarkli(sayi) == False:
        sayi = str(random.randint(pow(10,n-1), pow(10,n) - 1))
    return sayi
    
def tahminkontrol(sayi):
    kontrol = True
    
    if basamak(sayi) == False:
        print(inp, " basamaklı bir sayı girin!")
        kontrol = False
    if rakamlarfarkli(sayi) == False:
        print("Rakamlar farklı olmalı!")
        kontrol = False
    if rakamsayi(sayi) == False:
        print("Tahmininiz rakamlardan oluşmalı!")
        kontrol = False
    if ilkrakamsifir(sayi) == False:
        print("İlk rakam 0 olamaz!")
        kontrol = False
    return kontrol

def skorla(sayi, tahmin):
    artilar = 0
    eksiler = 0

    for i in tahmin:
        if i in sayi:
            if sayi.index(i) == tahmin.index(i):
                artilar += 1
            else:
                eksiler += 1
    if artilar == len(sayi):
        print("Kazandın!")
    else:
        print(" + ",artilar,"\n","- ",eksiler)
    return artilar

def degerlendir(i):
    if i <4:
        print("Şanlısın!")
    elif i <6:
        print("Dahi olmalısın!")
    elif i<7:
        print("Sen bu işi biliyorsun!")
    elif i<9:
        print("Fena değil!")
    elif i<11:
        print("Biraz kafa yorman gerek!")
    else:
        print("Sallaya sallaya tutturdun sonunda!")
    

def oyun(inp):
    skor=0
    i=0
   
    if inp not in range(3,10):
        while(inp not in range(3,10)):
            inp = int(input("Yanlış basamak bilgisi girdiniz. Kaç basamaklı bir sayıyı tahmin etmek istiyorsun (3-9 arası): "))
    sayi = sayiuret(inp)
    while skor != len(sayi):
        kontrol = False
        while kontrol == False:
            tahmin = input("\nSayı Girin: ")
            kontrol = tahminkontrol(tahmin)
        skor = skorla(sayi,tahmin)
        i += 1
    degerlendir(i)
    print(i,"denemede kazandınız!")
oyun(inp)

Kaç basamaklı bir sayıyı tahmin etmek istiyorsun (3-9 arası): 5

Sayı Girin: 123
5  basamaklı bir sayı girin!

Sayı Girin: 123456
5  basamaklı bir sayı girin!

Sayı Girin: 50245
Rakamlar farklı olmalı!

Sayı Girin: 12345
 +  0 
 -  2

Sayı Girin: 14568
 +  1 
 -  1

Sayı Girin: 54896
 +  0 
 -  2

Sayı Girin: 67942
 +  1 
 -  2

Sayı Girin: 10687
 +  0 
 -  3

Sayı Girin: 36875
 +  0 
 -  3


KeyboardInterrupt: ignored

<br><br><b>Problem 7.</b><br>(20 Puan)<br><br>
Aşağıda adam asmaca oyununun kodları verilmiştir.
Oyunu aşağıdaki gibi geliştirin.<br>
<b>a.</b> Kullanıcıdan kaç harfli kelime çözmek istediğini sorun ve ona göre kelime seçin.<br>
<b>b.</b> Kullanıcıya kaç hak istediğini sorun ve istediği kadar hak ile oynamasını sağlayın.<br>
<b>c.</b> Kullanıcıdan aldığınız cevap aşağıdaki şartları taşımıyorsa uyarı verip yeni tahmin girmesini isteyin. <br>(Bu hatalar ile karşılaşınca kullanıcının hakkından eksilme yapılacak.)<br>
   <b>c1.</b> Girilen tahmin tek karakter olmalı. (örnek hatalı girdiler: "ab", "/b*", "123", "kelime")<br>
   <b>c2.</b> Girilen tahmin harf olmalı. (örnek hatalı girdiler: "1", "*")<br>
   <b>c3.</b> Girilen karakter kullanıcı tarafından daha önce girilmemiş olmalı (Girilen her tahmin daha önce girilen tahminlerden farklı harf olmalı, farklı değilse hata vermeli)

In [ ]:
import random
metin ="Dünya, Güneş Sistemi'nde Güneş'e en yakın üçüncü gezegen olup şu an için üzerinde yaşam ve sıvı su barındırdığı bilinen tek astronomik cisimdir. Radyometrik tarihleme ve diğer kanıtlara göre 4,5 milyar yıldan fazla süre önce oluşmuştur. Dünya'nın yer çekimi, uzaydaki diğer nesnelerle, özellikle Güneş'le ve tek doğal uydusu Ay'la etkileşime girer. Dünya'nın Güneş'in etrafındaki yörüngesi, 365,256 güneş gün, yani bir yıldız yılı sürer. Bu süre içerisinde Dünya, kendi ekseni etrafında 366,265 kez döner."
metin_listesi = metin.split()
def kelime_uret(metin_listesi):
    tahmin_listesi = []
    for i in metin_listesi:
        #küçük harfe çevir
        i = i.lower()

        #Noktalama işaretlerini sil
        i = i.strip(",.:;/()?")

        #4 harften az olanları ele
        if (len(i) > 3) and (i not in tahmin_listesi) and (".,:;-" not in i):
            tahmin_listesi.append(i)
    """     
    #listeden rastgele eleman seçme
    k = random.randint(0,len(tahmin_listesi))
    kelime = tahmin_listesi[k]
    """
    
    #listeden rastgele eleman seçme
    kelime = random.choice(tahmin_listesi)
    
    return kelime

def adam_asmaca(metin_listesi = metin_listesi):
    hak_sayisi = 7
    aranilan_kelime = kelime_uret(metin_listesi)
    harf_sayisi = len(aranilan_kelime)
    #print(aranilan_kelime)

    print("Adam Asmaca Oyunu")
    print(harf_sayisi, "harfli kelime")
    print("!! ",hak_sayisi," hakkınız var !!")

    for i in range(0,harf_sayisi):
        print("_ ", end="")

    tutan_harfler=[]

    while True:
        #kullanıcıdan harf alalım
        tahmin = input("\nBir harf girin: ")
        tahmin = tahmin.lower()


        if tahmin in aranilan_kelime:
            tutan_harfler.append(tahmin)
        else:
            hak_sayisi = hak_sayisi -1

        kontrol = False
        for i in aranilan_kelime:
            if i in tutan_harfler:
                print(i, end=" ")
            else:
                print("_ ", end="")
                kontrol = True

        if kontrol == False:
            print("\nOYUNU KAZANDINIZ !!")
            break

        if hak_sayisi == 0:
            print("\n\nHakkınız bitti !!")
            print ("\n", aranilan_kelime)
            break

        print("\n!! ",hak_sayisi," hakkınız var !!")

In [ ]:
import random
metin ="Dünya, Güneş Sistemi'nde Güneş'e en yakın üçüncü gezegen olup şu an için üzerinde yaşam ve sıvı su barındırdığı bilinen tek astronomik cisimdir. Radyometrik tarihleme ve diğer kanıtlara göre 4,5 milyar yıldan fazla süre önce oluşmuştur. Dünya'nın yer çekimi, uzaydaki diğer nesnelerle, özellikle Güneş'le ve tek doğal uydusu Ay'la etkileşime girer. Dünya'nın Güneş'in etrafındaki yörüngesi, 365,256 güneş gün, yani bir yıldız yılı sürer. Bu süre içerisinde Dünya, kendi ekseni etrafında 366,265 kez döner."
metin_listesi = metin.split()
def kelime_uret(metin_listesi, harfSayi):
    tahmin_listesi = []
    for i in metin_listesi:
        #küçük harfe çevir
        i = i.lower()

        #Noktalama işaretlerini sil
        i = i.strip(",.:;/()?")

        #4 harften az olanları ele
        if (len(i) == harfSayi) and (i not in tahmin_listesi) and (".,:;-" not in i):
            tahmin_listesi.append(i)
    """     
    #listeden rastgele eleman seçme
    k = random.randint(0,len(tahmin_listesi))
    kelime = tahmin_listesi[k]
    """
    
    #listeden rastgele eleman seçme
    kelime = random.choice(tahmin_listesi)
    
    return kelime

def adam_asmaca(metin_listesi = metin_listesi):
    hak_sayisi = int(input("Kaç hak istersiniz: "))
    harf_sayisi = int(input("Kaç harfli bi kelime çözmek istersiniz: "))
    aranilan_kelime = kelime_uret(metin_listesi, harf_sayisi)
    
    #print(aranilan_kelime)

    print("Adam Asmaca Oyunu")
    print(harf_sayisi, "harfli kelime")
    print("!! ",hak_sayisi," hakkınız var !!")

    for i in range(0,harf_sayisi):
        print("_ ", end="")

    tutan_harfler=[]
    kullanilan_harfler = []
    while True:
        #kullanıcıdan harf alalım
        tahmin = input("\nBir harf girin: ")
        tahmin = tahmin.lower()
        x = 0
        while(x==0):
            if tahmin.isalpha() == False:
                hak_sayisi = hak_sayisi - 1
                break
                
            if tahmin in kullanilan_harfler:
                hak_sayisi = hak_sayisi - 1
                break
            
            if len(tahmin) > 1:
                hak_sayisi = hak_sayisi - 1
                break
            
            if tahmin in aranilan_kelime:
                tutan_harfler.append(tahmin)
            else:
                hak_sayisi = hak_sayisi - 1
            x = 1

        kontrol = False
        for i in aranilan_kelime:
            if i in tutan_harfler:
                print(i, end=" ")
            else:
                print("_ ", end="")
                kontrol = True

        if kontrol == False:
            print("\nOYUNU KAZANDINIZ !!")
            break

        if hak_sayisi == 0:
            print("\n\nHakkınız bitti !!")
            print ("\n", aranilan_kelime)
            break

        print("\n!! ",hak_sayisi," hakkınız var !!")
        kullanilan_harfler.append(tahmin)
adam_asmaca(metin_listesi)

Kaç hak istersiniz: 4
Kaç harfli bi kelime çözmek istersiniz: 5
Adam Asmaca Oyunu
5 harfli kelime
!!  4  hakkınız var !!
_ _ _ _ _ 
Bir harf girin: l
_ _ _ _ _ 
!!  3  hakkınız var !!

Bir harf girin: a
_ _ _ _ _ 
!!  2  hakkınız var !!

Bir harf girin: e
_ _ _ e _ 
!!  2  hakkınız var !!

Bir harf girin: y
_ _ _ e _ 
!!  1  hakkınız var !!

Bir harf girin: s
s _ _ e _ 
!!  1  hakkınız var !!

Bir harf girin: k
s _ _ e _ 

Hakkınız bitti !!

 sürer
